### PR TITLE
[ENP-820] Blackbox Exporter Enhanced

### PR DESCRIPTION
**Pull request by ArieDeha**
### Prometheus Blackbox Exporter on **dope** !

**Focused Files:**
- main.go
- config/config.go
- probe/icmp_qos.go
- probe/icmp_qos_test.go

Other files mostly are cleaning up “Unhandled error” and “whitespaces”

**DoD:**
Forked from prometheus blackbox exporter

**Add QoS Feature**
Create Test and Coverage for QoS

**ScreenShots:**
- Coverage Test 93%
image
- Results
image

**Commits**:
- [[ENP-820] Forked Prometheus Blackbox Exporter; Add QoS Features; Add Test and Coverage](https://github.com/devnetmonk/blackbox_exporter/commit/0085d3f262bd81fc6863f7eead6bde2f6913dcc3)
- [merge conflict: from latest prometheus](https://github.com/devnetmonk/blackbox_exporter/commit/989af2d979f2f988c41dffe19361f97f655c7c54)
- [cleanups and tidying go mod](https://github.com/devnetmonk/blackbox_exporter/commit/e6ba56a5e2ee39aed37164d556e4d9166171067c)

### CODE REVIEW

File: prober/icmp_qos.go [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+		probeQoSPacketLossLossCount = int64(s.PacketsSent - s.PacketsRecv)
+
+		count = int64(pinger.Count)
+
+		//// below line are added for making up the missing packets
+		////
+		totalLatency += (time.Duration(probeQoSPacketLossLossCount) * time.Microsecond) * pinger.Timeout
+
+		// microsecond to millisecond
+		if probeQoSPacketLossReceived > 0 {
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: mas aried, pengambilan min, max, avg stddev latency ngga dari result pingernya aja langsung ? 
https://github.com/prometheus-community/pro-bing/blob/v0.3.0/ping.go#L258

File: prober/icmp_qos.go [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+			jitter = -jitter
+		}
+		jitterSum += jitter
+		prevTime = float64(pinger.Timeout.Nanoseconds()) / perThousand
+		jitterMax = prevTime
+	}
+
+	pinger.OnRecv = func(pkt *ping.Packet) {
+		// jitter (in microseconds)
+		jitter = prevTime - float64(pkt.Rtt.Nanoseconds())/perThousand
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: ini kebalik ngga si mas ? kalo perhitungan pertama bukannya hasilnya jadi negative ?

File: prober/icmp_qos.go [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+	}, []string{"error_code", "error_message", "error_value"})
+
+	// Registering OpenMetrics Data
+	var probeQoSDurationSeconds = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "probe_qos_duration_seconds",
+		Help: "Total Durations of the pinger executions (in seconds)",
+	})
+
+	var probeQosLatencyGauge = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "probe_qos_latency_gauge",
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: metrics name disesuaikan dengan yang ada di RFC

File: prober/icmp_qos.go [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+		Help: "Probe QoS Latency Gauge (all are in milliseconds)",
+	}, []string{"aggregate"})
+	for _, lv := range []string{"total", "min", "max", "avg", "standard_deviation"} {
+		probeQosLatencyGauge.WithLabelValues(lv)
+	}
+
+	var probeQoSPacketLoss = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "probe_qos_packet_loss",
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: ini juga metrics name disesuaikan dengan yang ada di RFC

File: prober/icmp_qos.go [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+		prevTime  float64
+		jitter    float64
+		jitterSum float64
+		jitterMax float64
+		jitterMin float64
+
+		// request counter
+		count int64
+		// probe_duration_seconds
+		probeQoSDurationSecondsSecond float64
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: variable namingnya menurutku better langsung probeQoSDurationSeconds dengan perubahan variable naming metrics pake suffix metrics, atau  langsung probeQoSDurationSecondsValue. cmiiw

File: prober/icmp_qos.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+			"jitters", probeQoSJitterµs,
+			"jitter_max", probeQoSJitterGaugeMaxDiff,
+			"jitter_min", probeQoSJitterGaugeMinDiff,
+			"jitter_total", probeQoSJitterGaugeTotalDiff,
+		)
+
+		// probe_duration_seconds
+		endDuration = time.Now()
+		totalDuration = endDuration.Sub(startDuration)
+		probeQoSDurationSecondsSecond = float64(totalDuration.Microseconds()) / (perThousand * perThousand)
```
Comment (adhimaswaskita) [inline prober/icmp_qos.go]: ini bisa langsung pake .Miliseconds / 1000

**Review by adhimaswaskita — CHANGES_REQUESTED** [review]

File: prober/icmp_qos.go (hunk 1 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: done, on next commit

File: prober/icmp_qos.go (hunk 2 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: done, on next commit, dgn catetan, in a way, akan dibalik juga di line berikutnya klo hasilnya negative... expected memang akan ada nilai negative...

File: prober/icmp_qos.go (hunk 3 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: done, salah baca di rfc nya... on next commit

File: prober/icmp_qos.go (hunk 4 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: done... kebalik sama yg diatas

File: prober/icmp_qos.go (hunk 5 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: di remove karena ambil lsg aja, ga jadi pakai converter. done on next commit

File: prober/icmp_qos.go (hunk 6 above)
Comment (ArieDeha) [reply prober/icmp_qos.go]: duration lsg pakai .Seconds() karena lsg, udh float64

Comment (ArieDeha) [reply prober/icmp_qos.go]: done on next commit

**Comment by ArieDeha** [comment]
@adhimaswaskita Updated on commit: https://github.com/ArieDeha/blackbox_exporter/commit/3dffca3e98f80dd5ebab14eb0563b5a332632c61

**Review by adhimaswaskita — CHANGES_REQUESTED** [review]
fmt.Printf diremove mas @ArieDeha , atau ada perlu di print ?

**Comment by ArieDeha** [comment]
@adhimaswaskita udah yakkk!!

**Review by adhimaswaskita — APPROVED** [review]

**Comment by ganiamri** [comment]
LGTM thanks mas @ArieDeha

**Pull request merged**